### PR TITLE
test(style): run bash scripts through shellcheck

### LINKED ISSUE #3347: Proposal chore(scripts): use shellcheck to verify shell scripts
http://www.shellcheck.net/about.html

an example:

```
aledbf:bin aledbf$ shellcheck test-integration-ec2.sh

In test-integration-ec2.sh line 11:
export THIS_DIR=$(cd $(dirname $0); pwd)
                     ^-- SC2046: Quote this to prevent word splitting.
                               ^-- SC2086: Double quote to prevent globbing and word splitting.


In test-integration-ec2.sh line 14:
source $THIS_DIR/test-setup.sh
       ^-- SC2086: Double quote to prevent globbing and word splitting.


In test-integration-ec2.sh line 25:
    aws cloudformation delete-stack --stack-name $STACK_NAME
                                                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In test-integration-ec2.sh line 26:
    python $DEIS_ROOT/tests/bin/route53-wildcard.py delete $DEIS_TEST_DOMAIN $ELB_DNS_NAME
           ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                           ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                             ^-- SC2086: Double quote to prevent globbing and word splitting.


In test-integration-ec2.sh line 151:
        | grep InService | wc -l)
          ^-- SC2126: Consider using grep -c instead of grep|wc.
......
```

### PR DESCRIPTION
**Pull request by Joshua-Anderson**
Notes:

1. Jenkins now needs to have shellcheck installed. I'm defering to @mboersma and @sgoings on this.
2. I made a **lot** of changes to get these tests to make things pass, so this needs to be well manually tested to make sure I didn't break anything.

closes #3347

### CODE REVIEW

**Comment by Joshua-Anderson** [comment]
ref #4040

**Comment by Joshua-Anderson** [comment]
BTW Shellcheck caught 2 major errors.

Typo:
```
#/bin/bash
```

> 1 arg not supported:
```
#!/usr/bin/env bash -e
```


Is there a standard about whether to use `#!/bin/bash` or `#!/usr/bin/env bash`

**Comment by bacongobbler** [comment]
it's a recommendation: http://unix.stackexchange.com/questions/29608/why-is-it-better-to-use-usr-bin-env-name-instead-of-path-to-name-as-my

**Comment by Joshua-Anderson** [comment]
Is this something that should be standardized throughout all the shellscripts? They are a mixture of both.

**Comment by bacongobbler** [comment]
Yes, we've had PRs in the past that have migrated to /usr/bin/env before: https://github.com/deis/deis/pull/2538

**Comment by mboersma** [comment]
Here's how I got current `shellcheck` installed on CI master and nodes:
```
# Use cabal (Haskell installer) to build and install ShellCheck
apt-get install -yq cabal-install
cabal update
cd /tmp
git clone --single-branch https://github.com/koalaman/shellcheck.git
cd shellcheck
cabal install --global
apt-get purge cabal-install
```

@Joshua-Anderson could you add those lines to [setup-node.sh](https://github.com/deis/deis/blob/master/tests/bin/setup-node.sh) as part of this PR?

**Comment by Joshua-Anderson** [comment]
@mboersma Done!

**Comment by Joshua-Anderson** [comment]
Rebased and made sure that `$EXTRA_AWS_CLI_ARGS` isn't quoted.

File: contrib/aws/provision-aws-cluster.sh [hunk 1]
```diff
@@ -102,10 +110,12 @@ until [ `wc -w <<< $INSTANCE_STATUSES` -eq $DEIS_NUM_INSTANCES ]; do
...
 
     if [ $COUNTER -ne 1 ]; then sleep $SLEEPTIME; fi
-    echo "Waiting for instances to pass initial health checks ($(expr 61 - $COUNTER)0s) ..."
+    echo "Waiting for instances to pass initial health checks ($((61 - COUNTER))0s) ..."
+    # Don't warn about EXTRA_AWS_CLI_ARGS being unquoted
+    # shellcheck disable=SC2086
     INSTANCE_STATUSES=$(aws ec2 describe-instance-status \
         --filters Name=instance-status.reachability,Values=passed \
-        --instance-ids $INSTANCE_IDS \
+        --instance-ids "$INSTANCE_IDS" \
```
Comment (mboersma) [inline contrib/aws/provision-aws-cluster.sh]: This shouldn't be quoted: it makes all the instance IDs appear as one arg, which confused the provisioning script:
```
A client error (InvalidInstanceID.Malformed) occurred when calling the DescribeInstanceStatus operation: Invalid id: "i-661ddcaf
i-d4809923
i-ca80993d"
```

**Comment by mboersma** [comment]
D'oh, I think my diagnosis was wrong. CI still gets the same error after your most recent change, so probably we mistakenly quoted `INSTANCE_IDS` previously in the script.

**Comment by Joshua-Anderson** [comment]
@mboersma I think I found the guilty line in the `test-integration-aws.sh` script.

File: controller/Makefile [hunk 2]
```diff
@@ -4,6 +4,8 @@ include ../includes.mk
 
 all: build run
 
+SHELL_SCRIPTS = build.sh $(wildcard bin/*) $(wildcard migrations/data/*.sh)
```
Comment (bacongobbler) [inline controller/Makefile]: Something easier and more dynamic is to use `$(shell find $PWD -name '*.sh')`. That way we won't have to update this if we introduce new directories. Lazy searching ftw

Comment (Joshua-Anderson) [inline controller/Makefile]: Not every shell script is `*.sh`. `bin/boot` for example.

Comment (Joshua-Anderson) [inline controller/Makefile]: I'd be happy to add that in addition to the others ones though.

Comment (bacongobbler) [inline controller/Makefile]: good point. It'll remove the wildcard for the etcd data migrations, at the very least :)

Comment (Joshua-Anderson) [inline controller/Makefile]: Your command won't work because `$PWD` is very fragile. So if I ran

Comment (Joshua-Anderson) [inline controller/Makefile]: Fixed `$(shell find "." -name '*.sh')`

Comment (bacongobbler) [inline controller/Makefile]: That works! I didn't know if `.` would demonstrate the same behaviour as $PWD. I thought Makefiles were smarter than that. Good catch!

**Comment by Joshua-Anderson** [comment]
Added shellcheck to mesos and used @bacongobbler's suggestion for searching for shell files.

**Comment by carmstrong** [comment]
Looks like this needs a rebase, @Joshua-Anderson :(

**Comment by Joshua-Anderson** [comment]
Rebased!

**Comment by Joshua-Anderson** [comment]
I'm going to break this into smaller chunks so that it is easier to review.

**Comment by carmstrong** [comment]
> I'm going to break this into smaller chunks so that it is easier to review.

:+1: Going to close this PR, then. One less thing "awaiting review".

**Pull request closed without merging**